js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx

```diff
@@ -0,0 +1,126 @@
+} from './plugins';
+import CLASSES from '../classes';
+import { VirtualContent } from './virtual_content';
+import { getElementHeight } from '../utils';
+import { DEFAULT_ROW_HEIGHT, calculateRowHeight, calculateItemHeights } from './utils';
+import { Row } from '../types';
+
+export const viewFunction = ({
+  onRowsScrollPositionChange, scrollPosition, onRowsViewContentRender,
+  rowHeight, itemHeights,
+}: VirtualScrolling): JSX.Element => (
+  <Fragment>
+    <ValueSetter type={SetRowsViewScrollPositionAction} value={onRowsScrollPositionChange} />
+    <ValueSetter type={SetRowsViewContentRenderAction} value={onRowsViewContentRender} />
+    <ValueSetter type={ScrollingPositionValue} value={scrollPosition} />
+    <ValueSetter type={RowHeightValue} value={rowHeight} />
+    <ValueSetter type={ItemHeightsValue} value={itemHeights} />
+    <GetterExtender type={ViewportSkipValue} order={0} value={CalculateViewportSkipValue} />
+    <GetterExtender type={ViewportTakeValue} order={0} value={CalculateViewportTakeValue} />
+    <GetterExtender type={VisibleRows} order={1} value={AddLoadIndexToVisibleRows} />
+    <GetterExtender type={VisibleRows} order={2} value={CalculateVisibleRowsInViewport} />
+    <VirtualContent />
+  </Fragment>
+);
+
+@ComponentBindings()
+export class VirtualScrollingProps {
+  @OneWay()
+  mode: 'virtual' | 'infinite' = 'virtual';
+}
+
+@Component({
+  defaultOptionRules: null,
+  view: viewFunction,
+})
+export class VirtualScrolling extends JSXComponent(VirtualScrollingProps) {
+  @Consumer(PluginsContext)
+  plugins = new Plugins();
+
+  @InternalState()
+  scrollPosition = { top: 0, left: 0 };
+
+  @InternalState()
+  rowHeight = DEFAULT_ROW_HEIGHT;
+
+  @InternalState()
+  itemHeights: Record<number, number> = {};
+
+  @Mutable()
+  visibleRowHeights: number[] = [];
+
+  @Effect()
+  watchTotalCount(): () => void {
+    return this.plugins.watch(TotalCount, () => {
+      this.loadViewport();
+    });
+  }
+
+  onRowsScrollPositionChange(offset: ScrollOffset): void {
+    let { top, left } = offset;
+    top = Math.abs(top);
+    left = Math.abs(left);
+    if (this.scrollPosition.top !== top || this.scrollPosition.left !== left) {
+      this.scrollPosition = { top, left };
+      this.loadViewport();
+    }
+  }
+
+  loadViewport(): void {
+    let pageSize = this.plugins.getValue(PageSize) ?? 0;
+    pageSize = typeof pageSize !== 'number' ? 0 : pageSize;
+    const skip = this.plugins.getValue(ViewportSkipValue) ?? 0;
+    const take = this.plugins.getValue(ViewportTakeValue) ?? 0;
+    const pageIndex = Math.floor(pageSize > 0 ? skip / pageSize : 0);
+    const pageOffset = pageIndex * pageSize;
+    const skipForCurrentPage = skip - pageOffset;
+    const loadPageCount = Math.ceil(pageSize > 0 ? (take + skipForCurrentPage) / pageSize : 0);
+
+    this.plugins.callAction(SetPageIndex, pageIndex);
+    this.plugins.callAction(SetLoadPageCount, loadPageCount);
+  }
```

Comment:
```suggestion
  loadViewport(): void {
    const pageSize = this.plugins.getValue(PageSize) ?? 'all';
    if (pageSize !== 'all') {
      const skip = this.plugins.getValue(ViewportSkipValue) ?? 0;
      const take = this.plugins.getValue(ViewportTakeValue) ?? 0;
      const pageIndex = Math.floor(skip / pageSize);
      const pageOffset = pageIndex * pageSize;
      const skipForCurrentPage = skip - pageOffset;
      const loadPageCount = Math.ceil((take + skipForCurrentPage) / pageSize);

      this.plugins.callAction(SetPageIndex, pageIndex);
      this.plugins.callAction(SetLoadPageCount, loadPageCount);
    }
  }
```

Reply:
I think it is better to recalculate pageIndex and loadPageCount each time the loadViewport is called. Especially if pageSize is changed from any number to "all".